README.md
# byte_bloc

ByteBloc simplify binary array parsing for C/C++ structures.

## Usage

# Import

```dart
import 'package:byte_bloc/byte_bloc.dart';
```

# C/C++ structure
Example of C/C++ structure from embeded device
```cpp
struct DeviceStructure
{
    u_int32_t type;
    char name[12];
    unsigned isActive        :1;
    unsigned flags           :8;
};
```
# Binary array
Binary array of the structure will look like
```cpp
[223, 0, 77, 121, 32, 68, 101, 118, 105, 99, 101, 0, 0, 0, 7, 0, 0, 0];
```

# Model
From dart side we can implement model with factory fromByteBloc and method toByteBloc
```dart
class DeviceStructure {
  static const int NAME_SIZE = 12;
  static const int FLAGS_SIZE = 8;
  final int id;
  final String name;
  final bool isActive;
  final int flags;

  DeviceStructure({this.id, this.name, this.isActive, this.flags});

  factory DeviceStructure.fromByteBloc(ByteBloc byteBloc) {
    final id = byteBloc.readUint16();
    final name = byteBloc.readString(NAME_SIZE);
    final bitBloc = byteBloc.readBitBloc32();
    final isActive = bitBloc.readBool();
    final flags = bitBloc.readBits(FLAGS_SIZE);
    return DeviceStructure(id: id, name: name, isActive: isActive, flags: flags);
  }

  ByteBloc toByteBloc() {
    final byteBloc = ByteBloc.empty()
      ..writeUint16(id)
      ..writeString(name, NAME_SIZE)
      ..writeBitBloc32(BitBloc()
        ..writeBool(isActive)
        ..writeBits(flags, FLAGS_SIZE)
      )
      ..normalize();
    return byteBloc;
  }
}
```

# Deserialize model
```dart
final binaryArray = [223, 0, 77, 121, 32, 68, 101, 118, 105, 99, 101, 0, 0, 0, 7, 0, 0, 0];
final binaryArrayByteBloc = ByteBloc(Uint8List.fromList(binaryArray));
final deviceStructure = DeviceStructure.fromByteBloc(binaryArrayByteBloc);
```
# Serialize model
```dart
final resultByteBloc = deviceStructure.toByteBloc();
print(resultByteBloc.list);
```

# TODO
- Write unit tests for ByteBloc write methods


CHANGELOG.md
## 0.2.3
* Fix readByteBloc with nested blocs

## 0.2.2
* Minor fixes

## 0.2.1
* Write test for appendUint8List

## 0.2.0
* Add toString
* Add appendUint8List

## 0.1.9
* Minor fixes

## 0.1.8
* Minor fixes

## 0.1.7
* Update ByteBloc.empty constructor with size

## 0.1.6
* Minor fixes

## 0.1.5
* Write examples for bits stricture

## 0.1.4
* Minor fixes

## 0.1.3
* Minor fixes

## 0.1.2
* Minor fixes

## 0.1.1
* Update readme

## 0.1.0
* Add doc comments to public methods

## 0.0.1
* Implement ByteBloc read / write functionslity


test/src/bit_bloc_test.dart
import 'package:byte_bloc/byte_bloc.dart';
import 'package:test/test.dart';

void main() {
  group("read bits", () {
    final bitBloc = BitBloc(284); // 100011100
    test('read bits without offset', () {
      bitBloc.cursor = 0;
      expect(bitBloc.readBits(1), 0);
      expect(bitBloc.readBits(1), 0);
      expect(bitBloc.readBits(2), 3);
      expect(bitBloc.readBits(4), 1);
    });

    test('read bits with offset', () {
      bitBloc.cursor = 0;
      expect(bitBloc.readBits(2, offset: 2), 3);
      expect(bitBloc.readBits(5, offset: 4), 17);
    });
  });

  group("write bits", () {
    final bitBloc = BitBloc(0); // 100011100
    test('write bits without offset', () {
      bitBloc.cursor = 0;
      bitBloc.writeBits(0, 1);
      bitBloc.writeBits(0, 1);
      bitBloc.writeBits(3, 2);
      bitBloc.writeBits(17, 5);
      expect(bitBloc.bits, 284);
    });

    test('write bits with offset', () {
      bitBloc.cursor = 0;
      bitBloc.writeBits(3, 2, offset: 2);
      bitBloc.writeBits(1, 1);
      bitBloc.writeBits(1, 1, offset: 8);
      expect(bitBloc.bits, 284);
    });
  });
}


test/src/byte_bloc_test.dart
import 'dart:typed_data';

import 'package:byte_bloc/byte_bloc.dart';
import 'package:convert/convert.dart';
import 'dart:math' as math;

import 'package:test/test.dart';

void main() {
  group("read data", () {
    group("read uint", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x01, 0x02]));
      test('uint8', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint8(), 8);
        expect(byteBloc.readUint8(offset: 4), 1);
      });
      test('uint16', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint16(), 0x0108);
        expect(byteBloc.readUint16(offset: 2), 0x00ff);
      });
      test('uint32', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint32(), 0x00ff0108);
        expect(byteBloc.readUint32(), 0x0808ff01);
        expect(byteBloc.readUint32(offset: 4), 0x0808ff01);
      });

      test('uint64', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint64(), 0x0808ff0100ff0108);  
        expect(byteBloc.readUint64(offset: 2), 0x02010808ff0100ff);  
      });
    });
    group("read int", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x01, 0x02]));
      test('int8', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt8(), 8);
        expect(byteBloc.readInt8(offset: 4), 1);
      });
      test('int16', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt16(), 0x0108);
        expect(byteBloc.readInt16(offset: 2), 0x00ff);
      });
      test('int32', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt32(), 0x00ff0108);
        expect(byteBloc.readInt32(), 0x0808ff01);
        expect(byteBloc.readInt32(offset: 4), 0x0808ff01);
      });

      test('int64', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt64(), 0x0808ff0100ff0108);  
        expect(byteBloc.readInt64(offset: 2), 0x02010808ff0100ff);  
      });
    });
    group("read float", () {
      final byteBloc = ByteBloc(Uint8List.fromList(hex.decode("5c 8f cd 43 ec 91 cd 43 00 80 cd 43 00 80 cd 43".replaceAll(' ', ''))));
      test('float32', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readFloat32().roundFloat(), 411.12);
        expect(byteBloc.readFloat32().roundFloat(), 411.14);
      });

      test('float64', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readFloat64().roundFloat(), 922337203685477.6);
        expect(byteBloc.readFloat64(offset: 3).roundFloat(), -922337203685477.6);  
      });
    });
    group("read uint list", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08]));
      test('uint8List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint8List(2), [8, 1]);
        expect(byteBloc.readUint8List(2, offset: 4), [1, 255]);
      });
      test('uint16List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint16List(2), [0x0108, 0x00ff]);
        expect(byteBloc.readUint16List(2, offset: 2), [0x00ff, 0xff01]);
      });
      test('uint32List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint32List(2), [0x00ff0108, 0x0808ff01]);
      });

      test('uint64List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint64List(2), [0x0808ff0100ff0108, 0x0808ff0100ff0108]);  
      });
    });

    group("read int list", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08]));
      test('int8List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt8List(2), [8, 1]);
        expect(byteBloc.readInt8List(2, offset: 4), [1, -1]);
      });
      test('int16List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt16List(2), [0x0108, 0x00ff]);
        expect(byteBloc.readInt16List(2, offset: 2), [0x00ff, -255]);
      });
      test('int32List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt32List(2), [0x00ff0108, 0x0808ff01]);
      });

      test('int64List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt64List(2), [0x0808ff0100ff0108, 0x0808ff0100ff0108]);  
      });
    });

    group("read float list", () {
      final byteBloc = ByteBloc(Uint8List.fromList(hex.decode("5c 8f cd 43 ec 91 cd 43 00 80 cd 43 00 80 cd 43 5c 8f cd 43 ec 91 cd 43 00 80 cd 43 00 80 cd 43".replaceAll(' ', ''))));
      test('float32List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readFloat32List(2).map((e) => e.roundFloat()), [411.12, 411.14]);
      });

      test('float64List', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readFloat64List(2).map((e) => e.roundFloat()), [922337203685477.6, 922337203685477.6]);
      });
    });

    group("read uint square list", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08]));
      test('uint8SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint8SquareList(columns: 2, rows: 2), [[8, 1], [255, 0]]);
      });
      test('uint16SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint16SquareList(columns: 1, rows: 2), [[0x0108, 0x00ff]]);
      });
      test('uint32SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint32SquareList(columns: 1, rows: 2), [[0x00ff0108, 0x0808ff01]]);
      });

      test('uint64SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readUint64SquareList(columns: 1, rows: 2), [[0x0808ff0100ff0108, 0x0808ff0100ff0108]]);  
      });
    });

    group("read int square list", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08]));
      test('int8SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt8SquareList(columns: 1, rows: 2), [[8, 1]]);
        expect(byteBloc.readInt8SquareList(columns: 1, rows: 2, offset: 4), [[1, -1]]);
      });
      test('int16SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt16SquareList(columns: 1, rows: 2), [[0x0108, 0x00ff]]);
        expect(byteBloc.readInt16SquareList(columns: 1, rows: 2, offset: 2), [[0x00ff, -255]]);
      });
      test('int32SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt32SquareList(columns: 1, rows: 2), [[0x00ff0108, 0x0808ff01]]);
      });

      test('int64SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readInt64SquareList(columns: 1, rows: 2), [[0x0808ff0100ff0108, 0x0808ff0100ff0108]]);  
      });
    });

    group("read float square list", () {
      final byteBloc = ByteBloc(Uint8List.fromList(hex.decode("5c 8f cd 43 ec 91 cd 43 00 80 cd 43 00 80 cd 43 5c 8f cd 43 ec 91 cd 43 00 80 cd 43 00 80 cd 43".replaceAll(' ', ''))));
      test('float32SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readFloat32SquareList(columns: 1, rows: 2).map((e1) => e1.map((e2) => e2.roundFloat())), [[411.12, 411.14]]);
      });

      test('float64SquareList', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readFloat64SquareList(columns: 1, rows: 2).map((e1) => e1.map((e2) => e2.roundFloat())), [[922337203685477.6, 922337203685477.6]]);
      });
    });

    group("read bit bloc", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x01, 0x02]));
      test('bitBloc32', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readBitBloc32().bits, 0x00ff0108);
        expect(byteBloc.readBitBloc32().bits, 0x0808ff01);
        expect(byteBloc.readBitBloc32(offset: 4).bits, 0x0808ff01);
      });

      test('bitBloc64', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readBitBloc64().bits, 0x0808ff0100ff0108);  
        expect(byteBloc.readBitBloc64(offset: 2).bits, 0x02010808ff0100ff);  
      });
    });


    group("read byte bloc", () {
      final byteBloc = ByteBloc(Uint8List.fromList([0x08, 0x01, 0xff, 0, 0x01, 0xff, 0x08, 0x08, 0x01, 0x02]));
      test('byteBloc', () {
        byteBloc.cursor = 0;
        expect(byteBloc.readByteBloc(4, offset: 2).list, [0xff, 0, 0x01, 0xff]);
      });      
    });

    group("read nested byte bloc", () {
      final byteBloc = ByteBloc(Uint8List.fromList([174, 1, 6, 0, 0, 0, 1, 0, 221, 111, 16, 14, 0, 0]));
      test('byteBloc', () {
        final bloc = byteBloc.readByteBloc(12, offset: 2);
        expect(bloc.readUint32(), 6);
        expect(bloc.readByteBloc(4).list, [1, 0, 221, 111]);
        expect(bloc.readByteBloc(4).list, [16, 14, 0, 0]);
      });      
    });




  });

  group("append data", () {
    test('appendUint8List', () {
        final byteBloc = ByteBloc(Uint8List.fromList([1, 2, 3]));
        byteBloc.appendUint8List(Uint8List.fromList([4, 5, 6]));
        expect(byteBloc.list, [1, 2, 3, 4, 5, 6]);
      });
  });

  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}

final mapReduceInt = (it) => it.reduce((int value, int element) => 2);

extension FloatRoundExt on double {
  double roundFloat([int decimals = 4]) {
   int fac = math.pow(10, decimals);
   return (this * fac).round() / fac; 
  }
}

lib/src/bit_bloc.dart
import 'dart:math' as math;

class BitBloc {
  int bits;
  int cursor = 0;
  BitBloc([this.bits = 0]);

  /// Read bits
  int readBits(int size, {int offset}) {
    final byteOffset = offset ?? cursor;
    cursor = byteOffset + size;
    final mask = (((math.pow(2, size) as int) - 1) << byteOffset);
    final withMask = (bits & mask);
    final result = withMask >> byteOffset;
    return result;
  }

  /// Read bool
  bool readBool({int offset}) {
    final result = readBits(1, offset: offset);
    return result == 1;
  }

  /// Write bits
  void writeBits(int value, int size, {int offset}) {
    final byteOffset = offset ?? cursor;
    cursor = byteOffset + size;
    final maskedValue = value & ((math.pow(2, size) as int) - 1);
    final maskedValueWithOffset = maskedValue << byteOffset;
    final bitMask = ~(((math.pow(2, size) as int) - 1) << byteOffset);
    bits = (bits & bitMask) | maskedValueWithOffset;
  }

  /// Write bool
  void writeBool(bool value, {int offset}) {
    writeBits(value ? 1 : 0, 1, offset: offset);
  }
}


example/example.dart
import 'dart:typed_data';

import 'package:byte_bloc/byte_bloc.dart';

class DeviceStructure {
  static const int NAME_SIZE = 12;
  static const int FLAGS_SIZE = 8;
  final int id;
  final String name;
  final bool isActive;
  final int flags;

  DeviceStructure({this.id, this.name, this.isActive, this.flags});

  factory DeviceStructure.fromByteBloc(ByteBloc byteBloc) {
    final id = byteBloc.readUint16();
    final name = byteBloc.readString(NAME_SIZE);
    final bitBloc = byteBloc.readBitBloc32();
    final isActive = bitBloc.readBool();
    final flags = bitBloc.readBits(FLAGS_SIZE);
    return DeviceStructure(
      id: id,
      name: name,
      isActive: isActive,
      flags: flags,
    );
  }

  ByteBloc toByteBloc() {
    final byteBloc = ByteBloc.empty()
      ..writeUint16(id)
      ..writeString(name, NAME_SIZE)
      ..writeBitBloc32(BitBloc()
        ..writeBool(isActive)
        ..writeBits(flags, FLAGS_SIZE)
      )
      ..normalize();
    return byteBloc;
  }
}

main() async {
  final binaryArray = [223, 0, 77, 121, 32, 68, 101, 118, 105, 99, 101, 0, 0, 0, 7, 0, 0, 0];
  print(binaryArray);
  final binaryArrayByteBloc = ByteBloc(Uint8List.fromList(binaryArray));
  final deviceStructure = DeviceStructure.fromByteBloc(binaryArrayByteBloc);
  final resultByteBloc = deviceStructure.toByteBloc();
  print(resultByteBloc.list);
  assert(binaryArray.join(',') == resultByteBloc.list.join(','));
  print('Source and generated binary array are equal');
}